### ISSUE TITLE
Internal Server Error

### ISSUE THREAD

**Issue by Singularity9971** [kermitt2/grobid#112]
Hey so I'm running grobid on my Mac as a rest service and on a batch of about 400 documents, a couple of them have this error (file attached).
[error.txt](https://github.com/kermitt2/grobid/files/407876/error.txt)

When I do those documents separately from my browser they turn out fine. I have a python script which uses curl to call the service on a bunch of documents. Using the latest stable version of grobid.

**Comment by kermitt2** [comment]
Hello and thanks for the report ! 
Let's try to reproduce the error on our side... 
If doable - could you give us the log related to the error? I would be interested in the information related to the thread pool (number of engines in pool active/max). 
The python loop where the curl call appears would be helpful too, if you could also share this part I think it will be helpful.

**Comment by Singularity9971** [comment]
Hi I don't think I can get the log file, but here is the python code I used. I used multithreading in the script to speed up the process but bringing down the number of threads didn't change the errors.

subprocess.call(["curl",
                     "-v",
                     "-include",
                     "--form",
                     "input=@" + inFilename,
                     "localhost:" + port + "/processFulltextDocument",
                     "-o",
                     outFilename])

**Comment by kermitt2** [comment]
Thanks, the log are normally in a file called ```grobid.log```, by default at the place where you launch the server. 

What is the number of threads available on your machine and the amount of memory used by your java application server (for instance if jetty is used by maven with the command line: ```echo $MAVEN_OPTS```)?

For the python stuff, I think I would need the piece of code where the multithreading is visible - so the loop. 

Just a detail, in the curl call, the parameters "-v" and "-include" are not useful beyond testing and could be removed.

**Comment by Singularity9971** [comment]
Here is a portion of the log file with the exception.
[grobid.txt](https://github.com/kermitt2/grobid/files/411874/grobid.txt)

My machine has 4 cores, so the maximum I did was 8 threads since the cores support hyper threading.

Could you help me on how to get the memory used by the java application server? The command did not output anything.

Here is the python code:
[grobidcode.txt](https://github.com/kermitt2/grobid/files/411888/grobidcode.txt)

Thanks a lot for the help.

**Comment by Singularity9971** [comment]
I set my MAVEN_OPTS to -Xms1024m -Xmx4096m -XX:PermSize=1024m but the error is still there with the same documents.

**Comment by kermitt2** [comment]
Thanks ! 
And another question - if you reduce the number of threads used to 1 in the python script, do you still have this error?

**Comment by Singularity9971** [comment]
Yes reducing it to 1 fixes the error but obviously the execution time increases more than twice but for now this works for me, thanks!

**Comment by kermitt2** [comment]
Well it's not really a solution ;) The multithreading should work all the time. But it helps to understand. 

Apparently the problem is when all the threads are used and a pdf2xml timeout occurs - or maybe the pdf2xml timeout is due to the absence of an additional available thread (strange). I try to reproduce this situation on a mac to see if I get also an error.

If you set the number of threads used in the Python script to 2, do you have also the error? 

Thanks a lot for the help and your very fast answers!

**Comment by Singularity9971** [comment]
Haha I know, but since my dataset is small the increased time cost is something I can totally manage.

Interestingly, using 2 threads gives an error only on one document.

**Comment by lfoppiano** [comment]
@Singularity9971 I'm closing this, since it's very old. In case you're still interested @Singularity9971 , I believe the support to Mac OS has substantially improved over the last years. 🙂
